nncf/quantization/algorithms/weight_compression/lora_correction.py

```diff
@@ -0,0 +1,244 @@
+        nncf_logger.debug("Quantization noise through the correction process is saved to: ", losses_path)
+        df.to_csv(losses_path)
+
+        for name in df.columns:
+            plt.plot(df[name])
+            filename = name.replace("/", "_") + ".jpg"
+            plt.savefig(layer_dir / filename)
+            plt.clf()
+
+        delta = df.iloc[0] - df.iloc[-1]
+        nncf_logger.debug("Is quantization noise reduced for all layers: ", all(delta > 0))
+
+        _, ax = plt.subplots(1)
+        ax.plot(delta)
+        ax.set_xticklabels([])
+        delta_path = dump_dir / "qnoise_change.jpg"
+        nncf_logger.debug("Saving change in quantization noise for each layer to: ", delta_path)
+        plt.savefig(delta_path)
+        plt.clf()
+
+
+class LoraCorrectionAlgorithm:
+    """
+    Contains implementation of LoRA Correction algorithm.
+
+    The method reduces quantization noise after weight compression using low rank adapters.
+    """
+
+    def __init__(self, activations: Dict[str, List[Tensor]], lora_correction_params: AdvancedLoraCorrectionParameters):
+        """
+        :param activations: The input activations of the layers considered for compression.
+        :param lora_correction_params: parameters to configure the algorithm.
+        """
+        self._activations = activations
+        self._lora_correction_params = lora_correction_params
+        self._debug_interface = DebugInterface() if is_debug() else None
+
+    @property
+    def is_int8_adapters(self) -> bool:
+        return self._lora_correction_params.is_int8_adapters
+
+    def is_applicable(self, wc_params: WeightCompressionParameters):
+        return wc_params.compression_config.num_bits == 4
+
+    def calculate_adapters(
+        self, weight: Tensor, compressed_weight: Tensor, wc_params: WeightCompressionParameters
+    ) -> Tuple[Tensor, Tensor]:
+        """
+        Calculates low rank matrices for a given original and compressed weights.
+
+        :param weight: original floating-point weight matrix.
+        :param compressed_weight: compressed weight matrix.
+        :param wc_params: parameters of weight compression.
+        :return: two low rank matrices in the order of execution of corresponding linear layers.
+        """
+        try:
+            adapters = self.calculate_adapters_fn(
+                weight,
+                compressed_weight,
+                wc_params,
+                self._lora_correction_params,
+                self._activations,
+                self._debug_interface,
+            )
+        finally:
+            if self._debug_interface is not None:
+                self._debug_interface.dump_data()
+        return adapters
+
+    @staticmethod
+    def calculate_adapters_fn(
+        weight: Tensor,
+        compressed_weight: Tensor,
+        wc_params: WeightCompressionParameters,
+        lora_correction_params: AdvancedLoraCorrectionParameters,
+        activations: Dict[str, List[Tensor]],
```

Comment:
I would suggest to feed only those activations that the function needs for calculations.